kernel_tracing/metric_evaluation/test.py

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env python3
+
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright(c) 2022 Intel Corporation. All rights reserved.
+
+"""
+This script takes as an argument a path to a json file that contains test cases
+These cases specify a bpftrace script to collect data while specified shell commands are run
+When the shell commands finish, the bpftrace script is stopped and variables are collected
+The conditions in the spec file are evaled, replacing @vars with the values from the bpftrace script
+If any condition fails this script exits with code 1, otherwise it exits with code 0
+Shell commands are run in the directory of the spec file
+The bpftrace script path is relative to the spec file as well
+See spec.json for an example
+"""
+
+import os
+import sys
+import json
+import subprocess
+import re
+
+class colors:
+  BLUE = "\033[94m"
+  GREEN = "\033[92m"
+  RED = "\033[91m"
+  END = "\033[0m"
+def print_color(color, s):
+  print(color + s + colors.END)
+
+class BPF:
+  def __init__(self, bpftrace_script):
+    self.script = bpftrace_script
+    self.process = None
+
+  def start(self):
+    self.process = subprocess.Popen(["bpftrace", self.script], stdout=subprocess.PIPE, encoding="utf-8")
+    # Wait for bpftrace script to load
+    while True:
+      line = self.process.stdout.readline()
+      if "Attaching" in line: break
+
+  def stop(self):
+    # Send ctrl-c
+    self.process.send_signal(subprocess.signal.SIGINT)
+    self.process.wait()
+
+    # Collect printed variables for condition evaluation
+    lines = self.process.stdout.read().split("\n")
+    vars = {}
+    for line in lines:
+      if re.match("@.*: .*", line):
+        key, val = line.split(": ", 1)
+        vars[key] = val
+    return vars
+
+def check_condition(condition, vars):
+  for key, val in vars.items():
+    condition = condition.replace(key, val)
+  return eval(condition)
+
+def run_test(test_case, working_dir):
+  bpftrace_script = os.path.join(working_dir, test_case['script'])
+  shell_commands = test_case['shell']
+  conditions = test_case['conditions']
+  if isinstance(shell_commands, str):
+    shell_commands = [shell_commands]
+
+  # Start bpftrace collection
+  bpf = BPF(bpftrace_script)
+  bpf.start()
+
+  # Execute shell commands
+  for cmd in shell_commands:
+    print_color(colors.BLUE, f"$ {cmd}")
+    subprocess.call(cmd, cwd=working_dir, shell=True)
```

Comment:
https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module
> The recommended approach to invoking subprocesses is to use the [run()](https://docs.python.org/3/library/subprocess.html#subprocess.run) function for all use cases it can handle. For more advanced use cases, the underlying [Popen](https://docs.python.org/3/library/subprocess.html#subprocess.Popen) interface can be used directly.

https://docs.python.org/3/library/subprocess.html#older-high-level-api
> Prior to Python 3.5, these three functions comprised the high level API to subprocess. You can now use [run()](https://docs.python.org/3/library/subprocess.html#subprocess.run) in many cases, but lots of existing code calls these functions.
> `subprocess.call(args, *, ...`